solidity/contracts/bridge/Wallets.sol

```diff
@@ -305,6 +327,145 @@ library Wallets {
         emit NewWalletRegistered(ecdsaWalletID, walletPubKeyHash);
     }
 
+    /// @notice Handles a notification about a wallet heartbeat failure and
+    ///         triggers the wallet closure process.
+    /// @param publicKeyX Wallet's public key's X coordinate.
+    /// @param publicKeyY Wallet's public key's Y coordinate.
+    /// @dev Requirements:
+    ///      - The only caller authorized to call this function is `registry`
+    ///      - Wallet must be in Live state
+    function notifyWalletHeartbeatFailed(
+        Data storage self,
+        bytes32 publicKeyX,
+        bytes32 publicKeyY
+    ) external {
+        require(
+            msg.sender == address(self.registry),
+            "Caller is not the ECDSA Wallet Registry"
+        );
+
+        // Compress wallet's public key and calculate Bitcoin's hash160 of it.
+        bytes20 walletPubKeyHash = bytes20(
+            EcdsaLib.compressPublicKey(publicKeyX, publicKeyY).hash160()
+        );
+
+        requestWalletClosure(self, walletPubKeyHash);
+    }
+
+    /// @notice Handles a notification about a wallet action timeout
+    ///         (e.g. redemption request timeout) and requests slashing
+    ///         of the wallet operators. Triggers wallet closure process
+    ///         only if the wallet is still in the Live state. That means
+    ///         multiple action timeouts can be reported for the same
+    ///         wallet but only the first report will trigger wallet closure
+    ///         process.
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet
+    /// @dev Requirements:
+    ///      - Wallet must be in a state other than Unknown
+    function notifyWalletActionTimedOut(
+        Data storage self,
+        bytes20 walletPubKeyHash
+    ) external {
+        // TODO: Perform slashing of wallet operators.
+
+        WalletState walletState = self
+            .registeredWallets[walletPubKeyHash]
+            .state;
+
+        require(walletState != WalletState.Unknown, "Unknown wallet");
```

Comment:
Why don't we limit it only to `Live` and `MovingFunds` wallets?

Reply:
Actually, this is another argument for renaming `requestWalletClosure` to `moveFunds`. See that the wallet that is `MovingFunds` can still have a redemption timeout and we should only request to `moveFunds` when wallet is in `Live` state.

Reply:
What we discussed with @lukasz-zimnoch out of band:

![image](https://user-images.githubusercontent.com/4712360/160139458-6c579ec5-6213-4e55-87f4-a31fd6310521.png)

Only LIVE and MOVING_FUNDS wallet can be slashed.
MOVING_FUNDS wallet can have redemption timeouts reported and be slashed for that.
MOVING_FUNDS can not go to CLOSED until all redemptions are completed.

The last point prevents us from slashing the wallet that ended in `CLOSED` or `TERMINATED` states. These two are truly last possible states and no on-chain activity is possible for these wallets.

Reply:
tl;dr; we should require the state to be LIVE or MOVING_FUNDS to call this function.

Reply:
See https://github.com/keep-network/tbtc-v2/pull/176/commits/14dd3972d432744920b63f447bcb51361d4dd7ba